### PR TITLE
Fix error handling propagation

### PR DESCRIPTION
**Pull request by cgillum**
Resolves https://github.com/microsoft/durabletask-mssql/issues/161

For whatever reason, we forgot to fully implement exception propagation in this MSSQL backend for the Durable Task Framework. Fundamentally, the reason this is easy to miss is because the MSSQL provided doesn't serialize history events, but rather manually reads and writes them into structured SQL tables.

In this PR, we implement handling for both `ErrorPropagationMode.SerializeExceptions` (which we call "legacy") and `ErrorPropagationMode.UseFailureDetails` (which I'm treating as the new default).

To summarize the changes in this PR, we:
* Make sure to read/write failure details to and from the database as JSON
* Make sure we surface error details in the right way in the public API
* Add new tests or improve existing tests to verify that we flow exception details in expected ways

### CODE REVIEW

File: src/DurableTask.SqlServer/DTUtils.cs [hunk 1]
```diff
@@ -180,5 +200,52 @@ public static bool HasPayload(HistoryEvent e)
...
             };
         }
+
+        public static bool TryDeserializeFailureDetails(string jsonText, out FailureDetails? failureDetails)
+        {
+            try
+            {
+                failureDetails = DeserializeFromJson<FailureDetails>(jsonText);
+
+                // Depending on how the TaskHubWorker is configured, the text might be a FailureDetails JSON object
```
Comment (bachuv) [inline src/DurableTask.SqlServer/DTUtils.cs:210]: Just curious, can you share a link to the code where the `ErrorPropagationMode` is set to the default of `ErrorPropagationMode.UseFailureDetails`?

**Review by bachuv — APPROVED** [review]
Looks good to me! Just left a small question.

File: src/DurableTask.SqlServer/DTUtils.cs (hunk 1 above)
Comment (cgillum) [reply src/DurableTask.SqlServer/DTUtils.cs:210]: In this project, the default is set up in TestService.cs, line 68. The default mode can be set in the constructor parameter. For Orchestrations.cs, we don't specify anything, which causes the default value to be `UseFailureDetails`. Only in LegacyErrorPropagation.cs do we use the `SerializeExceptions` mode.

FWIW, the property is defined here: [`TaskHubWorker.ErrorPropagationMode`](https://github.com/Azure/durabletask/blob/6dc7134ceee4cec9073b86d5bb70da0cd4ca6731/src/DurableTask.Core/TaskHubWorker.cs#L130-L145). We don't set the default value explicitly, but enum properties default to zero, which in this case is [`SerializeExceptions`](https://github.com/Azure/durabletask/blob/6dc7134ceee4cec9073b86d5bb70da0cd4ca6731/src/DurableTask.Core/ErrorPropagationMode.cs#L28). This is the default because that was always the original behavior of the Durable Task Framework and changing the default behavior could break in-flight orchestrations for existing apps. It was much later that we added support for `UseFailureDetails`.

Comment (cgillum) [reply src/DurableTask.SqlServer/DTUtils.cs:210]: Not sure if this is what you were asking, but we also set it in the .NET Isolated SDK [here](https://github.com/microsoft/durabletask-dotnet/blob/e2f329258d1e02a762911bd89274a9a584b6ec04/src/Worker/Grpc/GrpcDurableTaskWorker.Processor.cs#L213-L217).

Java assumes this behavior as well, but doesn't require any explicit configuration since Java doesn't rely on the .NET implementation of `TaskHubWorker`. Basically, this is the _only_ option for Java.

Comment (bachuv) [reply src/DurableTask.SqlServer/DTUtils.cs:210]: Thanks for the explanation! I understand it a lot better now.

**Comment by aaditya8** [comment]
Hi, We are waiting for this fix from so long and upgraded sqlserver libraries and core libraries and started testing with both the modes "ErrorPropagationMode.SerializeExceptions", "ErrorPropagationMode.UseFailureDetails". Please find my observations help me with my use case. 

**Use Case :** 

We have one parent orchestration , one sub orchestration under parent orchestration , one task(Activity) under sub orchestration. 
I am expecting to propagate the exception occurred in sub orchestration task to both Sub orchestration and Parent orchestration. Below are my findings.

**Scenarios :**

1. With SerializeExceptions : Inner Exception is always null in Sub Orchestration and Parent Orchestration.
2. With UseFailureDetails : I am able to see the Failure details(Including stack trace) of the Sub Orchestration Activity in Sub Orchestration. Same way i am able to see the failure details of Sub Orchestration in Parent Orchestration but not able to see the Failure details of Sub orchestration activity. Its coming as null. 

Could you please help me with this.

**Comment by cgillum** [comment]
@aaditya8 please open a new issue for this and include details such as the kind of exceptions that are being raised. In that new issue, please also share code that shows how you're setting up your `TaskHubWorker`.
